add basic updating functionality

Davbra98 opened the pull request:
This is a very basic updater for Spytify. As outlined in #343 an updater for the application would be a nice feature. I tried to implement a very basic one here. I added an update project to the solution.

**How it works:**
1. When the user clicks on the ok button in the update messagebox or the symbol at the top the newest Github release URL is fetched by using the Github API and then the zip file is downloaded.
2. The updater is called with the path of Spytify as parameter and Spytify is closed.
3. The updater extracts the zip file overwriting the existing files.
4. If the updater was successfull Spytify is started with a parameter for the settings migration. If not an error message is shown. The updater also gets closed after that.
5. The settings are migrated. (Not tested now since the Github release does not contain the settings migration code by now)

**What is not completely finished by now:**
- I'm really bad at designing GUIs so that would need a rework.

**How to build:**
To get the build working make sure the updater project is building into Spytifys path in a subfolder Updater.
![grafik](https://user-images.githubusercontent.com/16635620/116001507-17374200-a5f5-11eb-8353-93b22e5ff94f.png)

close #343

jwallet commented on EspionSpotify.Updater/Update.cs:
```diff
@@ -0,0 +1,64 @@
+        }
+        public void extract(string path)
+        {
+            ZipFile zip = ZipFile.Read(path+"update.zip");
```
It may be not there yet?
It also occurs for some users with recorded files, so I don't know if a downloaded file can still be locked / in use by the process that downloaded it.

Davbra98 replied:
The whole update process is run on a single thread and executed synchronusly. So the extracting will not be started until the downlad was finished and the file exists.

jwallet reviewed (commented):
Hi, thanks for your pull request, at first sight I have some concern about the updater
- A console program instead of winform would have been quicker in my opinion, if you still want to go that way I suggest a splash screen style with just a Loader since you did not add internalization
- If I click the updater.exe I expect it to check the current app version, fetch any update, extract it and run it, delete the previous one in appdata, here it looks like it needs to be started from spytify to do that. I think it should be kind of a standalone app.

Other than that I left some comments

Davbra98 commented on EspionSpotify/GitHub.cs:
```diff
@@ -78,18 +98,44 @@ public static async Task GetVersion()
+            string json = new StreamReader(ws.GetResponseStream()).ReadToEnd();
+            json = json.Remove(0, 1);
+            json = json.Remove(json.Length - 1, 1);
+            string downlaodUrl = JObject.Parse(json)["browser_download_url"].ToString();
```
You are right. I missed it.

Davbra98 commented on EspionSpotify/GitHub.cs:
```diff
@@ -78,18 +98,44 @@ public static async Task GetVersion()
+            json = json.Remove(json.Length - 1, 1);
+            string downlaodUrl = JObject.Parse(json)["browser_download_url"].ToString();
+            WebClient wc = new WebClient();
+            wc.DownloadFile(downlaodUrl, "update.zip");
```
I have now added an exception handler for that and an error message. I think there is nothing elese we can do here

Davbra98 commented on EspionSpotify/GitHub.cs:
```diff
@@ -78,18 +98,44 @@ public static async Task GetVersion()
+            WebResponse ws = wr.GetResponse();
+            string json = new StreamReader(ws.GetResponseStream()).ReadToEnd();
+            json = json.Remove(0, 1);
+            json = json.Remove(json.Length - 1, 1);
```
I tried with a regex to make sure a version is included. Let me know if you had something different in mind.

Davbra98 commented:
I have now changed a lot of stuff. The updater can now be run standalone and also downloads the update file. I also switched to a console application. The only thing I don't know is if I can run Settings.Upgrade from the standalone Updater. So erverything but settings upgrading and deleting the old settings is now moved into the updater.

jwallet commented:
Ok thanks, it looks good, I will try to find some time to pull your branch and test it out, I may push commits before merging it.

jwallet commented on EspionSpotify.Updater/Program.cs:
```diff
@@ -0,0 +1,137 @@
+                    return true;
+                }
+                
+                if (currentVersion.StartsWith(m.Groups[1].Value))
```
official download should always start with spytify and end with zip "Spytify-v1.8.1.zip", so we want to ignore those:
`ADVERTISEMENT-ONLY-RECORDER-VERSION.-Spytify-v.1.8.zip`
`Spytify-v.1.8-debug.zip`
and version will always have major, but other are optional, so what I mean is:
`v.MAJOR.MINOR.PATCH` -- 4th number BUILD is always ignored, so it won't appear
so maybe a strict regex test will be better

```
var r = new System.Text.RegularExpressions.Regex(@"^Spytify(-|\.)(v?)(\.)?\d+(\.\d+)?(\.\d+)?.zip$", System.Text.RegularExpressions.RegexOptions.IgnoreCase);

Console.WriteLine(r.IsMatch("ADVERTISEMENT-ONLY-RECORDER-VERSION.-Spytify-v.1.8.zip")); // false
Console.WriteLine(r.IsMatch("Spytify-v.1.8-debug.zip")); // false
Console.WriteLine(r.IsMatch("Spytify.va.b.zip")); // false

Console.WriteLine(r.IsMatch("Spytify.v1.zip")); // true - with dot version

Console.WriteLine(r.IsMatch("Spytify-v1.zip")); // true - with dash version no-dot

Console.WriteLine(r.IsMatch("Spytify-v.1.zip")); // true - major - default (dash version dot)
Console.WriteLine(r.IsMatch("Spytify-v.1.8.zip")); // true - major minor
Console.WriteLine(r.IsMatch("Spytify-v.1.8.1.zip")); // true - major minor patch

Console.WriteLine(r.IsMatch("spytify-v.1.zip")); // true - lowercase
```

jwallet commented on EspionSpotify.Updater/Program.cs:
```diff
@@ -0,0 +1,137 @@
+                Asset targetRelease = null;
+                foreach (Asset asset in release.assets)
+                {
+                    if (Regex.IsMatch(asset.name, @"v\.((?:\d+\.)*\d+?).zip$"))
```
official download will always start with spytify and end with zip "Spytify-v1.8.1.zip", so we want to ignore those:
`ADVERTISEMENT-ONLY-RECORDER-VERSION.-Spytify-v.1.8.zip`
`Spytify-v.1.8-debug.zip`
and version will always have major, but other are optional, so what I mean is:
`v.MAJOR.MINOR.PATCH` -- 4th number BUILD is always ignored, so it won't appear
so maybe a strict regex test will be better

```
var r = new System.Text.RegularExpressions.Regex(@"^Spytify(-|\.)(v?)(\.)?\d+(\.\d+)?(\.\d+)?.zip$", System.Text.RegularExpressions.RegexOptions.IgnoreCase);

Console.WriteLine(r.IsMatch("ADVERTISEMENT-ONLY-RECORDER-VERSION.-Spytify-v.1.8.zip")); // false
Console.WriteLine(r.IsMatch("Spytify-v.1.8-debug.zip")); // false
Console.WriteLine(r.IsMatch("Spytify.va.b.zip")); // false

Console.WriteLine(r.IsMatch("Spytify-v.1.zip")); // true - major
Console.WriteLine(r.IsMatch("Spytify-v.1.8.zip")); // true - major minor
Console.WriteLine(r.IsMatch("Spytify-v.1.8.1.zip")); // true - major minor patch

Console.WriteLine(r.IsMatch("spytify-v.1.zip")); // true - lowercase

Console.WriteLine(r.IsMatch("Spytify.v1.zip")); // true - with dot version

Console.WriteLine(r.IsMatch("Spytify-v1.zip")); // true - with dash version no-dot
```

Davbra98 replied:
Ok
I changed the regex now to the one you suggested

jwallet commented:
I assigned myself on the issue just to prevent another pull request, I was unable to assign the issue to you since you did not comment on it. however, it does not really matter
